feat: Use ForwardPassCallback api from TRTLLM to register end of forward pass callback to enable cuda graphs

nv-kmcgill53 opened the pull request:
#### Overview:
Updating the KVBM connector worker to make use of the [trtllm ForwardPassCallback api](https://github.com/NVIDIA/TensorRT-LLM/pull/8275). This api allows KVBM to register a CUDA event and a callback for when that event is triggered at the end of the forward pass. Using this API allows KVBM to be used with or without cuda graphs enabled in TRTLLM. 

The way this API is implemented in the KVBM Connector worker, we hand TRTLLM a ForwardPassCallback object with an event we describe. TRTLLM, at the end of the forward pass, will (optionally if defined) insert this event into the forward pass stream and immediately call the callback provided. This allows KVBM to wait on this event being triggered to schedule the block offloading. 


<!-- Describe your pull request here. Please read the text below the line, and make sure you follow the checklist.-->

#### Details:

Addresses https://github.com/ai-dynamo/dynamo/issues/3387

<!-- Describe the changes made in this PR. -->

#### Where should the reviewer start?

<!-- call out specific files that should be looked at closely -->

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

- closes GitHub issue: #xxx

nv-kmcgill53 committed:
Added enum for toggle between how we offload cache after the forward pass

By default we count the layerwise executions in the forward pass
but in the case of using cuda graphs with trtllm, forward passes
two and greater don`t trigger our kvbm counts. Thus, we hook into
wait_for_save() which is called at the end of every forward pass,
bypassing the need to count the layers from the stream events.

wait_for_save() might be slower since when getting the stream of events
we only have to sync on the cuda events to know it is safe to
transfer. This can be optimized later.

nv-kmcgill53 committed:
test to prove the previous functionality holds

Added a test to prove trtllm worker + kvbm - cuda graphs
still work with the required changes.

nv-kmcgill53 committed:
Revert "Added enum for toggle between how we offload cache after the forward pass"

This reverts commit 0921d6b96519103fae5cc02bbdf1623886843286.

nv-kmcgill53 committed:
Added enums for the different offloading modes and added them to the kvbm configuration struct

nv-kmcgill53 committed:
fixing compile bugs

nv-kmcgill53 committed:
initial pass at api interface for passing a cuda event to trtllm

Signed-off-by: Kyle McGill <kmcgill@nvidia.com>

nv-kmcgill53 committed:
using new TRTLLM ForwardPassCallback api

nv-kmcgill53 committed:
clean up for pr

nv-kmcgill53 committed:
Revert "Added enums for the different offloading modes and added them to the kvbm configuration struct"

This reverts commit 3feb5cdb0b5cae326ab27927fa9d2f446309ca7c.

nv-kmcgill53 committed:
clean up for pr

nv-kmcgill53 committed:
whitespace

nv-kmcgill53 committed:
synchronizing on the cuda event instead of on the stream

nv-kmcgill53 committed:
Refactored the API to provide a callable to TRTLLM

Instead of returning a class which needs to have a
function implemented, the API now provides a callable
which will be called at the end of the forward pass.

nv-kmcgill53 committed:
Added feature to make DynamoKVBMConnectorWorker backward compatible

nv-kmcgill53 committed:
code rabbit suggestions

nv-kmcgill53 commented on tests/kvbm/test_cuda_graph.py:
```diff
@@ -0,0 +1,187 @@
+PROMPT = "In the heart of Eldoria, an ancient land of boundless magic and mysterious creatures, lies the long-forgotten city of Aeloria. Once a beacon of knowledge and power, Aeloria was buried beneath the shifting sands of time, lost to the world for centuries. You are an intrepid explorer, known for your unparalleled curiosity and courage, who has stumbled upon an ancient map hinting at ests th ...
+
+class DynamoWorkerProcess(ManagedProcess):
+    """Process manager for Dynamo worker with vLLM backend"""
```
where are the other places of vllm in this file? I only see this one.

ziqifan617 replied:
I remembered I saw in another place. but I checked now all gone. we are good

nv-kmcgill53 committed:
test file PR typo and formatting suggestions

nv-kmcgill53 committed:
adding test for with/without cuda graphs

nv-kmcgill53 committed:
fixing tests

nv-kmcgill53 commented on tests/kvbm/test_cuda_graph.py:
```diff
@@ -0,0 +1,187 @@
+        with DynamoWorkerProcess(request, "decode", engine_config_with_cuda_graph_and_kvbm) as worker:
+            logger.info(f"Worker PID: {worker.get_pid()}")
+
+            print(f"{send_completion_request(PROMPT, 100, timeout=10)}")
```
updated PR to test the response code of the response object.

nv-kmcgill53 commented on lib/bindings/python/rust/llm/block_manager/vllm/connector/trtllm_worker.rs:
```diff
@@ -211,17 +213,23 @@ impl Worker for KvConnectorWorker {
-            }
+            let _ = self.execute_offload_operations();
         }
         Ok(())
```
updated.

nv-kmcgill53 commented on tests/kvbm/test_cuda_graph.py:
```diff
@@ -0,0 +1,187 @@
+    This test verifies a TRTLLM worker is able to serve requests when 
+    cuda graphs are enabled in pytorch. KVBM should be able to offload 
+    blocks regardless.
+    """
```
updated doc strings

ziqifan617 commented on lib/bindings/python/src/dynamo/llm/trtllm_integration/connector/kvbm_connector_worker.py:
```diff
@@ -106,8 +125,10 @@ def save_kv_layer(self, layer_idx: int, stream: torch.cuda.Stream):
         """
-        self.events[layer_idx].record(stream)
-        self._connector.save_kv_layer(layer_idx)
+        # This is a WAR to not collide with registering a forward_pass_callable
```
not sure why this is called a `WAR to not collide with registering a forward_pass_callable`

my understanding is that this is just a normal branching logic

nv-kmcgill53 replied:
Ideally, there should only be one path taken for the offloading. The reason for this comment are my personal feelings around why this if-statement exists rather than any technical reason. Would you like me to remove this?

nv-kmcgill53 committed:
merge main into feature branch

nv-kmcgill53 committed:
Removing incorrect comment

nv-kmcgill53 committed:
added copyrights to test files

nv-kmcgill53 committed:
pre-commit formatting

nv-kmcgill53 committed:
adding trtllm mark so test does not run in github

nv-kmcgill53 committed:
disabling cuda graph tests due to incompatible apis in the current state

Cuda graph tests are disabled until dynamo is able to advance onto
trtllm 1.2+ because while the trtllm changes were developed on
trtllm 1.1.0rc5, the changes needed to be merged into main.
This caused the dynamo tests to require trtllm main => 1.2+
which has breaking api changes.

nv-kmcgill53 committed:
fix assertion in test

nv-kmcgill53 committed:
enabling tests for PR testing

nv-kmcgill53 committed:
removing bad logging messages from debugging

nv-kmcgill53 committed:
Merge branch 'main' into kmcgill/kvmb-cuda-graphs

nv-kmcgill53 committed:
adding correct markers and skipping tests

nv-kmcgill53 committed:
Merge branch 'main' into kmcgill/kvmb-cuda-graphs
